File: src/main/java/in/tazj/k8s/letsencrypt/util/DetectCloudPlatform.java
package in.tazj.k8s.letsencrypt.util;

import com.amazonaws.regions.Regions;

import org.xbill.DNS.Lookup;
import org.xbill.DNS.TextParseException;

import lombok.extern.slf4j.Slf4j;
import lombok.val;

/**
 * Inspects the environment to determine which cloud platform the controller is running on.
 *
 * Currently supported are Amazon Web Services and Google Cloud Platform.
 */
@Slf4j
public class DetectCloudPlatform {
  public enum CloudPlatform {
    GCP, AWS, UNKNOWN
  }

  public static CloudPlatform detectCloudPlatform() {
    log.info("Detecting current cloud platform ...");

    if (detectAmazonWebServices()) {
      log.info("Cloud platform is Amazon Web Services");
      return CloudPlatform.AWS;
    }

    if (detectGoogleCloudPlatform()) {
      log.info("Cloud platform is Google Cloud Platform");
      return CloudPlatform.GCP;
    }

    log.info("Could not detect cloud platform");
    return CloudPlatform.UNKNOWN;
  }

  /**
   * Attempts to figure out whether the environment is AWS by checking the current region.
   */
  private static boolean detectAmazonWebServices() {
    val region = Regions.getCurrentRegion();
    return (region != null);
  }

  /**
   * Attempts to figure out whether the environment is GCP by attempting to resolve the metadata
   * server host.
   */
  private static boolean detectGoogleCloudPlatform() {
    try {
      val lookup = new Lookup("metadata.google.internal");
      return lookup.run() != null;
    } catch (TextParseException e) {
      log.error("Error while detecting Google Cloud Platform: {}", e.getMessage());
      return false;
    }
  }
}

File: src/main/kotlin/in/tazj/k8s/letsencrypt/util/DetectCloudPlatform.kt
package `in`.tazj.k8s.letsencrypt.util

import com.amazonaws.regions.Regions
import org.slf4j.LoggerFactory
import org.xbill.DNS.Lookup

enum class CloudPlatform {
    GCP, AWS, UNKNOWN
}

/**
 * Inspects the environment to determine which cloud platform the controller is running on.
 *
 * Currently supported are Amazon Web Services and Google Cloud Platform.
 */
fun detectCloudPlatform(): CloudPlatform {
    val log = LoggerFactory.getLogger(CloudPlatform::class.java)
    log.info("Detecting current cloud platform ...")
    return when {
        detectAmazonWebServices() -> {
            log.info("Cloud platform is Amazon Web Services")
            CloudPlatform.AWS
        }
        detectGoogleCloudPlatform() -> {
            log.info("Cloud platform is Google Cloud Platform")
            CloudPlatform.GCP
        }
        else -> {
            log.warn("Could not detect cloud platform")
            CloudPlatform.UNKNOWN
        }
    }
}

private fun detectAmazonWebServices(): Boolean {
    val region = Regions.getCurrentRegion()
    return (region != null)
}

private fun detectGoogleCloudPlatform(): Boolean {
    val lookup = Lookup("metadata.google.internal")
    return lookup.run() != null
}
